### PR TITLE
Feature/sg 36 yolox loss

### PR DESCRIPTION
**Pull request by yurkovak**
Adding YOLOX loss function with:
* center sampling
* dynamic k-matching 

discussed in [the paper](https://arxiv.org/pdf/2107.08430.pdf) in 2.1 in sections **Multi positives** and **SimOTA**

**Notes:**
- The implementation is based on [the original code](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/yolox/models/yolo_head.py) with many added comments. I can totally add many more comments and docstrings.
- The input is expected in the same shapes as those produced by YoloV5, just for consistency, but for an anchorless model.

### CODE REVIEW

File: src/super_gradients/training/losses/yolox_loss.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+        if self.reduction == "mean":
+            loss = loss.mean()
+        elif self.reduction == "sum":
+            loss = loss.sum()
+
+        return loss
+
+
+def bboxes_iou(bboxes_a, bboxes_b, xyxy=True):
```
Comment (yurkovak) [inline src/super_gradients/training/losses/yolox_loss.py:53]: We already have super_gradients.training.utils.detection_utils.calculate_bbox_iou_elementwise, not sure we can interchangeable use it, but for now it's just a copy from the source

File: src/super_gradients/training/losses/yolox_loss.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                gt_bboxes_per_image = labels_im[:, 2:6] * self.im_size
+                gt_classes = labels_im[:, 1]
+                bboxes_preds_per_image = bbox_preds[batch_idx]
+
+                try:
+                    gt_matched_classes, fg_mask, pred_ious_this_matching, matched_gt_inds, num_fg_img = \
+                        self.get_assignments(batch_idx, num_gt, total_num_anchors, gt_bboxes_per_image,
+                                             gt_classes, bboxes_preds_per_image,
+                                             expanded_strides, x_shifts, y_shifts, cls_preds, obj_preds)
+                except RuntimeError:
```
Comment (yurkovak) [inline src/super_gradients/training/losses/yolox_loss.py:145]: Do we want to keep this? It's supposed to help when during training it goes OOM while trying to do the assignment. I actually saw it happen a few times

**Review by yurkovak — COMMENTED** [review]
Got a few questions

File: src/super_gradients/training/losses/yolox_loss.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/losses/yolox_loss.py:53]: let's add a TODO: comment to check whether its interchangeable

File: src/super_gradients/training/losses/yolox_loss.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    return area_i / (area_a[:, None] + area_b - area_i)
+
+
+class YoloXDetectionLoss(_Loss):
+    """
+    Calculate YOLO V5 loss:
+    L = L_objectivness + L_iou + L_classification (+ L_l1)
+    """
+
+    def __init__(self, strides, im_size, num_classes, use_l1=False, center_sampling_radius=2.5):
```
Comment (shaydeci) [inline src/super_gradients/training/losses/yolox_loss.py:80]: Please add doctoring documentation here, and for the rest of the methods as they are not that trivial...

**Review by shaydeci — CHANGES_REQUESTED** [review]

**Comment by shaydeci** [comment]
Im closing this PR along with the other ones related to yolo X integration as it has been stale for a while.
Let's reopen it once we continue working on yoloX.
